### ISSUE TITLE
[no-unnecessary-condition] allow unnecessary optional chaining in list of conditions for uniformity

### ISSUE THREAD

**Issue by snebjorn** [typescript-eslint/typescript-eslint#3100]
I would like to propose a new option for the `@typescript-eslint/no-unnecessary-condition` rule which would allow unnecessary optional chaining in if statements.

Lets say you've written
```ts
if (
  foo?.a &&
  foo?.b &&
  foo?.c
) {}
```
Then this rule would prompt you to remove `?.` on `b` and `c` because it's not necessary. And that is absolutely correct.
```ts
if (
  foo?.a &&
  foo.b &&
  foo.c
) {}
```
However to keep it easy to alter and transparent when skimming the code (in a code review), I'd like an option to turn this off for these types of expressions.

Lets say someone changed the order of the `foo`s because it wasn't sorted or changed from `&&` to `||` then the resulting diff touch more lines of code

examples
```diff
if (
- foo?.a &&
+ foo?.a ||
- foo.b &&
+ foo?.b &&
  foo.c
) {}
```
vs
```diff
if (
- foo?.a &&
+ foo?.a ||
  foo?.b &&
  foo?.c
) {}

**Comment by bradzacher** [comment]
It's not an easy thing to do, unfortunately.

From a human's perspective it's simple - if the chain has an optional chain on a member earlier in the logical, then don't report on it.

But from a static analysis perspective it's REALLY hard.
The types (what the rule is driven by) show that the optional chain is unnecessary, so we can't leverage those at all; meaning you can only do this via AST.
I won't get into it (try chucking your example into astexplorer.net to see the AST structure), but needless to say that there are a _lot_ of steps involved in doing correct comparisons.

I'll leave this open if someone wants to look into it - but the above stands as a warning.

**Comment by Josh-Cena** [comment]
It has been 3 years and this proposal hasn't really got much feedback. As mentioned this is not an easy thing to do, for example, is this "uniform"? `a?.b.c?.d && a?.b.c?.e` Or this? `a.b?.() && a.b?.cc` It's overall a very subjective thing to check and hard to formalize a rule for, so it's better that you just use a disable comment, or fork and extend this rule.
